File: server/core/src/main/scala/org/infinispan/server/core/LifecycleCallbacks.scala
package org.infinispan.server.core

import org.infinispan.lifecycle.AbstractModuleLifecycle
import org.infinispan.server.core.ExternalizerIds._
import org.infinispan.factories.{ComponentRegistry, GlobalComponentRegistry}
import org.infinispan.configuration.global.GlobalConfiguration
import org.infinispan.configuration.cache.Configuration
import org.infinispan.factories.components.ComponentMetadataRepo

/**
 * Module lifecycle callbacks implementation that enables module specific
 * {@link org.infinispan.marshall.AdvancedExternalizer} implementations to be registered.
 *
 * @author Galder Zamarreño
 * @since 5.0
 */
class LifecycleCallbacks extends AbstractModuleLifecycle {

   override def cacheManagerStarting(gcr: GlobalComponentRegistry, globalCfg: GlobalConfiguration) {
      LifecycleCallbacks.componentMetadataRepo = gcr.getComponentMetadataRepo
   }

   override def cacheStarting(cr: ComponentRegistry, cfg: Configuration, cacheName: String) {
      cfg.storeAsBinary().enabled(false)
   }

}

object LifecycleCallbacks {

   var componentMetadataRepo: ComponentMetadataRepo = _

}

File: lucene/lucene-directory/src/main/java/org/infinispan/lucene/LifecycleCallbacks.java
package org.infinispan.lucene;

import org.infinispan.configuration.global.GlobalConfiguration;
import org.infinispan.factories.GlobalComponentRegistry;
import org.infinispan.factories.annotations.InfinispanModule;
import org.infinispan.lifecycle.ModuleLifecycle;
import org.infinispan.lucene.impl.PersistenceContextInitializerImpl;
import org.infinispan.marshall.protostream.impl.SerializationContextRegistry;

/**
 * Module lifecycle callbacks implementation that enables module specific
 * {@link org.infinispan.commons.marshall.AdvancedExternalizer} implementations to be registered.
 *
 * @author Galder Zamarreño
 * @author Sanne Grinovero
 * @since 5.0
 */
@InfinispanModule(name = "lucene-directory", requiredModules = "core")
public class LifecycleCallbacks implements ModuleLifecycle {

   @Override
   public void cacheManagerStarting(GlobalComponentRegistry gcr, GlobalConfiguration globalCfg) {
      registerSerializationContext(gcr);
   }

   public static void registerSerializationContext(GlobalComponentRegistry gcr) {
      SerializationContextRegistry ctxRegistry = gcr.getComponent(SerializationContextRegistry.class);
      ctxRegistry.addContextInitializer(SerializationContextRegistry.MarshallerType.PERSISTENCE, new PersistenceContextInitializerImpl());
   }
}
